controllers/concept.go

```diff
@@ -148,3 +152,147 @@ func (u ConceptController) RetrieveBreakdownStatsBySourceIdAndCohortIdAndConcept
+	for _, breakdownStat := range breakdownStats {
+		concept_value := breakdownStat.ConceptValue
+		if concept_value == "" {
+			concept_value = "empty string"
+		}
+		concept_values_to_people_count[concept_value] = breakdownStat.NpersonsInCohortWithValue
+	}
+
+	return concept_values_to_people_count
+}
+
+func generateConceptRow(cohortName string, conceptToPeopleCount map[string]int, sortedConcepts []string) []string {
+	cohortSize := 0
+	for _, peopleCount := range conceptToPeopleCount {
+		cohortSize += peopleCount
+	}
+
+	row := []string{cohortName, strconv.Itoa(cohortSize)}
+	for _, concept := range sortedConcepts {
+		row = append(row, strconv.Itoa(conceptToPeopleCount[concept]))
+	}
+
+	return row
+}
+
+func (u ConceptController) getFilteredConceptRows(sourceId int, cohortId int, conceptIds []int, breakdownConceptId int, sortedConcepts []string) ([][]string, error) {
+	conceptIdToName := make(map[int]string)
+	conceptInformations, err := u.conceptModel.RetrieveInfoBySourceIdAndConceptIds(sourceId, conceptIds)
+	if err != nil {
+		return nil, fmt.Errorf("Could not retrieve concept informations due to error: %s", err.Error())
+	}
+	for _, conceptInformation := range conceptInformations {
+		conceptIdToName[conceptInformation.ConceptId] = conceptInformation.ConceptName
+	}
+
+	var rows [][]string
+	for idx, conceptId := range conceptIds {
+		filterConceptIds := conceptIds[0 : idx+1]
+		breakdownStats, err := u.conceptModel.RetrieveBreakdownStatsBySourceIdAndCohortIdAndConceptIds(sourceId, cohortId, filterConceptIds, breakdownConceptId)
+		if err != nil {
+			return nil, fmt.Errorf("Could not retrieve concept Breakdown for concepts %v due to error: %s", filterConceptIds, err.Error())
+		}
+
+		conceptValuesToPeopleCount := getConceptValueToPeopleCount(breakdownStats)
+		rows = append(rows, generateConceptRow(conceptIdToName[conceptId], conceptValuesToPeopleCount, sortedConcepts))
+	}
+
+	return rows, nil
+}
+
+func (u ConceptController) RetrieveAttritionTable(c *gin.Context) {
+	sourceId, cohortId, conceptIds, err1 := parseSourceIdAndCohortIdAndConceptIds(c)
+	breakdownConceptId, err2 := utils.ParseNumericArg(c, "breakdownconceptid")
+
+	if err1 == nil && err2 == nil {
+		breakdownStats, err := u.conceptModel.RetrieveBreakdownStatsBySourceIdAndCohortId(sourceId, cohortId, breakdownConceptId)
+		if err != nil {
+			c.JSON(http.StatusInternalServerError, gin.H{"message": "Error retrieving breakdown stats", "error": err})
+			c.Abort()
+			return
+		}
+		conceptValuesToPeopleCount := getConceptValueToPeopleCount(breakdownStats)
+
+		cohortName, err := getCohortName(cohortId)
+		if err != nil {
+			c.JSON(http.StatusInternalServerError, gin.H{"message": "Error retrieving cohort name", "error": err})
+			c.Abort()
+			return
+		}
+
+		headerAndNonFilteredRow := genereateHeaderAndNonFilteredRow(cohortName, conceptValuesToPeopleCount)
+		sortedConcepts := getSortedConcepts(conceptValuesToPeopleCount)
```

Comment:
I think these are concept _values_ and not plain concepts, right? To avoid confusion with concepts themselves, I would prefer more specific names like: `getSortedConceptValues` instead of `getSortedConcepts` here. Similarly, the variable name `sortedConcepts` would then be called `sortedConceptValues`, etc, etc.